File: backend/app/services/scraper/sources/hirehi_filters.py
"""HireHi listing filters — values taken from the site's own search UI/API."""

from __future__ import annotations

CATEGORIES = [
    ("development", "Разработка"),
    ("devops", "DevOps"),
    ("qa", "Тестирование"),
    ("analytics", "Аналитика"),
    ("design", "Дизайн"),
    ("management", "Менеджмент"),
    ("marketing", "Маркетинг"),
    ("sales", "Продажи"),
    ("finance", "Финансы"),
    ("recruiting", "Рекрутинг"),
]

SUBCATEGORIES = [
    ("ml_ai", "ML/AI"),
    ("python", "Python"),
    ("java", "Java"),
    ("backend", "Backend"),
    ("go", "Go"),
    ("data_engineer", "Data Engineer"),
    ("frontend", "Frontend"),
    ("fullstack", "Fullstack"),
    ("netc", ".NET/C#"),
    ("cpp", "C++"),
    ("php", "PHP"),
    ("nodejs", "Node.js"),
    ("kotlin", "Kotlin"),
    ("rust", "Rust"),
    ("mobile", "Mobile"),
    ("android", "Android"),
    ("ios", "iOS"),
    ("onec", "1C"),
    ("erp_crm", "ERP / CRM"),
]

FORMATS = [
    ("удалённо", "удалённо"),
    ("офис", "офис"),
    ("гибрид", "гибрид"),
    ("удалённо по РФ", "удалённо по РФ"),
]

LEVELS = [
    ("intern", "intern"),
    ("junior", "junior"),
    ("middle", "middle"),
    ("senior", "senior"),
    ("lead", "lead"),
    ("head", "head"),
]

ENGLISH = [
    ("english", "нужен английский"),
    ("no_english", "не нужен английский"),
]

CONTACTS = [
    ("direct_contact", "прямой контакт"),
    ("linkedin", "LinkedIn"),
    ("email", "Email"),
    ("telegram", "Telegram"),
]

SORTS = [
    ("date", "сначала новые"),
    ("salary_desc", "больше денег"),
]

_LIST_KEYS = (
    "format",
    "level",
    "subcategory",
    "english",
    "direct_contact",
    "salary",
    "country",
    "region",
    "industry",
)


def _as_list(value: object) -> list[str]:
    if value is None:
        return []
    if isinstance(value, str):
        return [value] if value else []
    if isinstance(value, list):
        return [str(item) for item in value if item]
    return []


def normalize_hirehi_params(raw: dict | None) -> dict:
    data = dict(raw or {})
    return {
        "category": data.get("category") or "development",
        "search": (data.get("search") or "").strip(),
        "sort": data.get("sort") or "date",
        "format": _as_list(data.get("format")),
        "level": _as_list(data.get("level")),
        "subcategory": _as_list(data.get("subcategory")),
        "english": _as_list(data.get("english")),
        "direct_contact": _as_list(data.get("direct_contact")),
        "salary": _as_list(data.get("salary")),
        "country": _as_list(data.get("country")),
        "region": _as_list(data.get("region")),
        "industry": _as_list(data.get("industry")),
        "city": (data.get("city") or "").strip(),
    }


def auto_name(params: dict) -> str:
    labels = dict(SUBCATEGORIES)
    parts: list[str] = []
    search = (params.get("search") or "").strip()
    if search:
        parts.append(search)
    for slug in params.get("subcategory") or []:
        label = labels.get(slug, slug)
        if label.lower() != search.lower():
            parts.append(label)
    parts.extend(params.get("format") or [])
    levels = params.get("level") or []
    if levels:
        parts.append(", ".join(levels))
    english = params.get("english") or []
    if "english" in english:
        parts.append("EN")
    if "no_english" in english:
        parts.append("без EN")
    unique = list(dict.fromkeys(parts))
    return " · ".join(unique[:5]) or "HireHi поиск"

File: frontend/lib/hirehi-filters.ts
export const CATEGORIES = [
  { value: "development", label: "Разработка" },
  { value: "devops", label: "DevOps" },
  { value: "qa", label: "Тестирование" },
  { value: "analytics", label: "Аналитика" },
  { value: "design", label: "Дизайн" },
  { value: "management", label: "Менеджмент" },
  { value: "marketing", label: "Маркетинг" },
  { value: "sales", label: "Продажи" },
  { value: "finance", label: "Финансы" },
  { value: "recruiting", label: "Рекрутинг" },
] as const;

export const SUBCATEGORIES = [
  { value: "ml_ai", label: "ML/AI" },
  { value: "python", label: "Python" },
  { value: "java", label: "Java" },
  { value: "backend", label: "Backend" },
  { value: "go", label: "Go" },
  { value: "data_engineer", label: "Data Engineer" },
  { value: "frontend", label: "Frontend" },
  { value: "fullstack", label: "Fullstack" },
  { value: "netc", label: ".NET/C#" },
  { value: "cpp", label: "C++" },
  { value: "php", label: "PHP" },
  { value: "nodejs", label: "Node.js" },
  { value: "kotlin", label: "Kotlin" },
  { value: "rust", label: "Rust" },
  { value: "mobile", label: "Mobile" },
  { value: "android", label: "Android" },
  { value: "ios", label: "iOS" },
  { value: "onec", label: "1C" },
  { value: "erp_crm", label: "ERP / CRM" },
] as const;

export const FORMATS = [
  { value: "удалённо", label: "удалённо" },
  { value: "офис", label: "офис" },
  { value: "гибрид", label: "гибрид" },
  { value: "удалённо по РФ", label: "удалённо по РФ" },
] as const;

export const LEVELS = [
  { value: "intern", label: "intern" },
  { value: "junior", label: "junior" },
  { value: "middle", label: "middle" },
  { value: "senior", label: "senior" },
  { value: "lead", label: "lead" },
  { value: "head", label: "head" },
] as const;

export const ENGLISH = [
  { value: "english", label: "нужен английский" },
  { value: "no_english", label: "не нужен английский" },
] as const;

export const CONTACTS = [
  { value: "direct_contact", label: "прямой контакт" },
] as const;

export const SORTS = [
  { value: "date", label: "сначала новые" },
  { value: "salary_desc", label: "больше денег" },
] as const;

export type HireHiFilters = {
  category: string;
  search: string;
  sort: string;
  format: string[];
  level: string[];
  subcategory: string[];
  english: string[];
  direct_contact: string[];
};

export const EMPTY_FILTERS: HireHiFilters = {
  category: "development",
  search: "",
  sort: "date",
  format: [],
  level: [],
  subcategory: [],
  english: [],
  direct_contact: [],
};

function asList(value: unknown): string[] {
  if (Array.isArray(value)) return value.map(String).filter(Boolean);
  if (typeof value === "string" && value) return [value];
  return [];
}

export function filtersFromParams(params: Record<string, unknown> | undefined): HireHiFilters {
  const data = params ?? {};
  return {
    category: String(data.category || "development"),
    search: String(data.search || ""),
    sort: String(data.sort || "date"),
    format: asList(data.format),
    level: asList(data.level),
    subcategory: asList(data.subcategory),
    english: asList(data.english),
    direct_contact: asList(data.direct_contact),
  };
}

export function autoName(filters: HireHiFilters): string {
  const subLabels = Object.fromEntries(SUBCATEGORIES.map((s) => [s.value, s.label]));
  const parts: string[] = [];
  const search = filters.search.trim();
  if (search) parts.push(search);
  for (const slug of filters.subcategory) {
    const label = subLabels[slug] ?? slug;
    if (label.toLowerCase() !== search.toLowerCase()) parts.push(label);
  }
  parts.push(...filters.format);
  if (filters.level.length) parts.push(filters.level.join(", "));
  if (filters.english.includes("english")) parts.push("EN");
  if (filters.english.includes("no_english")) parts.push("без EN");
  return [...new Set(parts)].slice(0, 5).join(" · ") || "HireHi поиск";
}

export function summarize(filters: HireHiFilters): string {
  const subLabels = Object.fromEntries(SUBCATEGORIES.map((s) => [s.value, s.label]));
  const bits = [
    filters.search.trim(),
    ...filters.subcategory.map((s) => subLabels[s] ?? s),
    ...filters.format,
    ...filters.level,
    ...ENGLISH.filter((e) => filters.english.includes(e.value)).map((e) => e.label),
    ...filters.direct_contact.map(() => "прямой контакт"),
  ].filter(Boolean);
  return [...new Set(bits)].join(" · ") || "без фильтров";
}

export function toggleValue(list: string[], value: string): string[] {
  return list.includes(value) ? list.filter((item) => item !== value) : [...list, value];
}
